cassiopeia/data/CassiopeiaTree.py

```diff
@@ -68,117 +73,291 @@ class CassiopeiaTree:
+        must have an entry for every node in the tree.
+
+        Args:
+            mapping: A mapping containing character state assignments for every
+                node
+        
+        Raises:
+            CassiopeiaTreeError if the tree is not initialized or if the
+                mapping does not contain assignments for every node.
+        """
+        if self.__network is None:
+            raise CassiopeiaTreeError("Tree has not been initialized.")
+
+        if len(np.setdiff1d(self.nodes, character_matrix.keys())) > 0:
+            raise CassiopeiaTreeError(
+                "Mapping does not account for all the nodes."
+            )
+
+        character_matrix = {}
+        for n in self.nodes:
+            if self.is_leaf():
+                character_matrix[n] = mapping[n]
+            self.__set_character_states(n, mapping[n])
+
+        character_matrix = pd.DataFrame.from_dict(
+            character_matrix, orient="index"
+        )
+        self.__original_character_matrix = character_matrix.copy()
+        self.__current_character_matrix = character_matrix.copy()
+
+    def get_original_character_matrix(self) -> pd.DataFrame:
+        """Gets the original character matrix.
+
+        The returned character matrix is the original character matrix of
+        observations. Downstream operations might change the character state
+        observations for the cells and if this happens, the changes will
+        not be reflected here. Instead, the changes will be reflected in the 
+        character matrix obtained with `get_current_character_matrix`.
+
+        Returns:
+            A copy of the original, unmodified character matrix.
+
+        Raises:
+            CassiopeiaTreeError if the character matrix does not exist. 
+        """
+        if self.__original_character_matrix is None:
+            raise CassiopeiaTreeError("Character matrix does not exist.")
+        return self.__original_character_matrix.copy()
+
+    def get_current_character_matrix(self) -> pd.DataFrame:
+        """Gets the original character matrix.
+
+        The returned character matrix is the modified character matrix of
+        observations. When downstream operations are used to change the
+        character state observations in the leaves of the tree, these changes
+        will be reflected here. A "raw" version of the character matrix can
+        be found in the `get_original_character_matrix` method.
+
+        Returns:
+            A copy of the modified character matrix.
+
+        Raises:
+            CassiopeiaTreeError if the character matrix does not exist. 
+        """
+        if self.__current_character_matrix is None:
+            raise CassiopeiaTreeError("Character matrix does not exist.")
+        return self.__current_character_matrix.copy()
 
     @property
     def n_cell(self) -> int:
-        """Returns number of cells in character matrix.
+        """Returns number of cells in tree.
```

Comment:
Maybe we should make it explicit in the docstring if the tree does not exist, it returns 0 (instead of throwing an error for example, which is done by the other properties)

Reply:
Yeah, this is a good observation, I also found the treatment of this border case (no tree and no matrix => 0 cells) a little bit weird. Probably better to raise an error?

Reply:
Probably a fair point.

I'll also throw an error if character states have not been initialized and no character matrix exists